core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,248 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ * -- Add note that in the case of cascading closure that listeners to the parent are always
+ * notified before listeners on the children.
+ *
+ */
+// Discussion point. Using java.io.Closeable here so we can inherit AutoCloseable where its available.
+public final class Context {
+
+  /**
+   * Provide an AutoCloseable (via Closeable) for use with try-with-resource blocks that will
+   * close a context and detach it.
+   */
+  public class ContextCloseable implements Closeable {
+    @Override
+    public void close() {
+      try {
+        Context.this.detach();
+      } finally {
+        Context.this.close(null);
+      }
+    }
+  }
+
+  /**
+   * The logical root context which is {@link #current()} if no other context is bound. This context
+   * is already closed and will not retain any listeners.
+   */
+  public static final Context ROOT = new Context();
+  static {
+    ROOT.close(null);
+  }
+
+  /**
+   * A listener notified on context {@link #close}.
+   */
+  public interface Listener {
```

Comment:
Close events still seem a little awkward in this interface in that which contexts should or should not be listened to isn't quite clear. Perhaps we should be more explicit.

Should we just make a separate interface `CloseableContext` (which defines `Listener` and `withListener`). We could then have all context instances implement it, except for ROOT. All `with` methods would return `CloseableContext`s.